{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "08f4a11d-c9c6-4412-b920-f5334e6aeb35",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "46353bbd-fbd9-4b12-82dd-f2bdab935c99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>query</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b4484.faa</td>\n",
       "      <td>b4484_alleles.fasta</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     subject                query\n",
       "0  b4484.faa  b4484_alleles.fasta"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_inputs_mapping = pd.read_csv('./blast_input_mapping.csv',\n",
    "            names=[\"subject\", \"query\"]\n",
    "           )\n",
    "blast_inputs_mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bd842bcd-3ad0-457c-90b5-cc4094aa7a88",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALLELE_FASTA_PATH = './data/allele_fastas/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6dbfaf4d-cebe-4012-9a09-e3444a95dfb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/allele_fastas/b4484_alleles.fasta\n"
     ]
    }
   ],
   "source": [
    "gene_allele_names_d = dict()\n",
    "for fasta_file in blast_inputs_mapping['query']:\n",
    "    gene_name = fasta_file.replace(\"_alleles.fasta\", '')\n",
    "    gene_allele_names_d[gene_name] = set()\n",
    "    fasta_path = ALLELE_FASTA_PATH + fasta_file\n",
    "    print(fasta_path)\n",
    "    with open(fasta_path, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('>'):\n",
    "                allele_name = line.replace('>','').rstrip(\"\\n\")\n",
    "                gene_allele_names_d[gene_name].add(allele_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2cd24dc4-d958-45bd-96c8-0b710089801a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "\n",
    "# TODO: should be using blast_inputs_mapping instead of what's in the ref_fastas dir\n",
    "gene_aa_len_d = dict()\n",
    "# for ref_fasta_file in glob.glob('./data/ref_fastas/*.faa'):\n",
    "for ref_fasta_file in blast_inputs_mapping['subject']:\n",
    "    gene_name = ref_fasta_file.replace('.faa','')\n",
    "    ref_fasta_file_path = './data/ref_fastas/' + ref_fasta_file\n",
    "    with open(ref_fasta_file_path, 'r') as file:\n",
    "        aa_seq = ''\n",
    "        for line in file:\n",
    "            if '>' not in line:\n",
    "                aa_seq += line.rstrip(\"\\n\")\n",
    "        gene_aa_len_d[gene_name] = len(aa_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9548c261-0c06-400c-811c-a197faa3dcb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# String parsing helper functions\n",
    "\n",
    "# fast find all chars in string: https://stackoverflow.com/questions/52452911/finding-all-positions-of-a-character-in-a-string\n",
    "def _find_all_idx(string, character):\n",
    "    idx = string.find(character)\n",
    "    while idx != -1:\n",
    "        yield idx\n",
    "        idx = string.find(character, idx + 1)\n",
    "        \n",
    "\n",
    "def _get_char_idxs_in_str(s, c):\n",
    "    char_idxs = list(_find_all_idx(s, c))\n",
    "    char_idxs = [x for x in char_idxs]\n",
    "    return char_idxs\n",
    "    \n",
    "\n",
    "CHAR_OF_INTEREST = '-'\n",
    "assert(_get_char_idxs_in_str('AR-D-G', CHAR_OF_INTEREST)==[2,4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9c91ad02-83a4-45bf-b362-442431e5c1cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# BLAST results glossary: https://www.ncbi.nlm.nih.gov/books/NBK62051/\n",
    "# Description of special characters on the match string: https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=BlastHelp\n",
    "# The databases alignments are displayed as pairs of matches between query and subject sequence.\n",
    "# A middle line between the query and subject sequence displays the status of a letter.\n",
    "# For protein alignments (e.g, BLASTP/BLASTX/TBLASTN), identities present the letter,\n",
    "# conservative substitutions present a \"+\", and nothing otherwise. This is the default view.\n",
    "# https://www.ncbi.nlm.nih.gov/books/NBK62051/: \"If the aligned residues have similar physico-chemical properties or have a\n",
    "# positive score in the governing scoring matrix the substitution is said to be conservative.\"\n",
    "# '-' in subject string indicate an insertion\n",
    "\n",
    "# BLAST subject and query results string are always the same length.\n",
    "\n",
    "\n",
    "# Won't find differences in the beginning of genes since doesn't compare query vs subject start and end match positions.\n",
    "# This logic was removed since alleles can have enormous differences on either of their sequences.\n",
    "\n",
    "# Best to keep the returned data seperate for finding\n",
    "# consecutive mutations according to AA positions.\n",
    "\n",
    "BLAST_INDEL_CHAR = '-'\n",
    "def _get_ins_d(subject_str, query_str, subject_match_start_pos):\n",
    "    ins_d = dict()\n",
    "    for idx in _get_char_idxs_in_str(subject_str, BLAST_INDEL_CHAR):\n",
    "        pos = idx+subject_match_start_pos\n",
    "        ins_d[pos] = query_str[idx]\n",
    "    return ins_d\n",
    "\n",
    "# Won't find differences in the beginning of genes since doesn't compare query vs subject start and end match positions.\n",
    "def _get_del_d(subject_str, query_str, subject_match_start_pos):\n",
    "    del_d = dict()\n",
    "    for idx in _get_char_idxs_in_str(query_str, BLAST_INDEL_CHAR):\n",
    "        pos = idx+subject_match_start_pos\n",
    "        del_d[pos] = subject_str[idx]\n",
    "    return del_d\n",
    "    \n",
    "\n",
    "sbjct = 'AR-D-G'  # insertion @ 3 and 5\n",
    "qry = '-RBDHG'  # delection @ 1\n",
    "assert(_get_ins_d(sbjct, qry, 1)=={3: 'B', 5: 'H'})\n",
    "assert(_get_ins_d(sbjct, qry, 2)=={4: 'B', 6: 'H'})  # test effects of subject_match_start_pos\n",
    "\n",
    "assert(_get_del_d(sbjct, qry, 1)=={1: 'A'})\n",
    "assert(_get_del_d(sbjct, qry, 2)=={2: 'A'})  # test effects of subject_match_start_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "00385c86-88d9-4b9f-8da3-b34533cbfbf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALL_AMINO_ACIDS = {'A','R','N','D','B','C','E','Q','Z','G','H','I','L','K','M','F','P','S','T','W','Y','V'}\n",
    "\n",
    "def _get_sub_aa_d(subject_str, match_str, query_str, indel_aa_pos_l, subject_match_start_pos):\n",
    "    seq_chng_d = dict()\n",
    "    \n",
    "    for c_i in range(0, len(match_str)):\n",
    "        if match_str[c_i] not in ALL_AMINO_ACIDS:  # finding BLAST characters indicating mismatch or gap\n",
    "            pos = subject_match_start_pos+c_i  # subject_match_start_pos+c_i since subject_match_start_pos represents the first charater and the index c_i == 0\n",
    "            if pos not in indel_aa_pos_l:\n",
    "                seq_chng_d[pos] = {'s':subject_str[c_i], 'q':query_str[c_i]}\n",
    "    \n",
    "    return seq_chng_d\n",
    "\n",
    "\n",
    "assert(_get_sub_aa_d('ARPGCQ', '+ PG Q', 'MTPGBQ', [], 1)=={1: {'s':'A', 'q':'M'}, 2: {'s':'R', 'q':'T'}, 5: {'s':'C', 'q':'B'}})\n",
    "assert(_get_sub_aa_d('ARPGCQ', '+ PG Q', 'MTPGBQ', [5], 1)=={1: {'s':'A', 'q':'M'}, 2: {'s':'R', 'q':'T'}})  # testing use of indel_aa_pos_l\n",
    "assert(_get_sub_aa_d('ARPGCQ', '+ PG Q', 'MTPGBQ', [], 2)=={2: {'s':'A', 'q':'M'}, 3: {'s':'R', 'q':'T'}, 6: {'s':'C', 'q':'B'}})  # test effects of subject_match_start_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cceb38fc-7c60-4395-bb40-1db6f4f36fec",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import groupby\n",
    "from operator import itemgetter\n",
    "\n",
    "\n",
    "def _get_consecutive_pos_l(pos_l):\n",
    "    cnsc_pos_l = []\n",
    "    for k, g in groupby(enumerate(pos_l), lambda ix : ix[0] - ix[1]):\n",
    "        cnsc_pos_l.append(list(map(itemgetter(1), g)))\n",
    "    return cnsc_pos_l\n",
    "\n",
    "\n",
    "mut_pos_l = [1, 4,5, 7,8,9, 22]\n",
    "assert(_get_consecutive_pos_l(mut_pos_l)==[[1], [4, 5], [7, 8, 9], [22]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ec894d6d-d3e4-42de-9015-4764ac918edb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>query</th>\n",
       "      <th>blast output</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b4484.faa</td>\n",
       "      <td>b4484_alleles.fasta</td>\n",
       "      <td>blast_b4484_alleles.xml</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     subject                query             blast output\n",
       "0  b4484.faa  b4484_alleles.fasta  blast_b4484_alleles.xml"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "blast_inputs_mapping['blast output'] = blast_inputs_mapping['query'].apply(lambda s: 'blast_' + s.replace('.fasta', '.xml'))\n",
    "blast_inputs_mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "91b612c3-def8-43a6-a070-c2c4f469ed73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA range</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>AA mutation type</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b4484</td>\n",
       "      <td>pangenome variant</td>\n",
       "      <td>(120, 120)</td>\n",
       "      <td>1</td>\n",
       "      <td>substitution</td>\n",
       "      <td>M</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b4484</td>\n",
       "      <td>pangenome variant</td>\n",
       "      <td>(105, 105)</td>\n",
       "      <td>1</td>\n",
       "      <td>substitution</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b4484</td>\n",
       "      <td>pangenome variant</td>\n",
       "      <td>(164, 164)</td>\n",
       "      <td>1</td>\n",
       "      <td>substitution</td>\n",
       "      <td>R</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b4484</td>\n",
       "      <td>pangenome variant</td>\n",
       "      <td>(11, 11)</td>\n",
       "      <td>1</td>\n",
       "      <td>substitution</td>\n",
       "      <td>S</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b4484</td>\n",
       "      <td>pangenome variant</td>\n",
       "      <td>(147, 147)</td>\n",
       "      <td>1</td>\n",
       "      <td>substitution</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b4484</td>\n",
       "      <td>pangenome variant</td>\n",
       "      <td>(125, 125)</td>\n",
       "      <td>1</td>\n",
       "      <td>substitution</td>\n",
       "      <td>T</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b4484</td>\n",
       "      <td>pangenome variant</td>\n",
       "      <td>(153, 153)</td>\n",
       "      <td>1</td>\n",
       "      <td>substitution</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b4484</td>\n",
       "      <td>pangenome variant</td>\n",
       "      <td>(27, 27)</td>\n",
       "      <td>1</td>\n",
       "      <td>substitution</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b4484</td>\n",
       "      <td>pangenome variant</td>\n",
       "      <td>(28, 28)</td>\n",
       "      <td>1</td>\n",
       "      <td>substitution</td>\n",
       "      <td>D</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b4484</td>\n",
       "      <td>pangenome variant</td>\n",
       "      <td>(31, 31)</td>\n",
       "      <td>1</td>\n",
       "      <td>substitution</td>\n",
       "      <td>H</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>145 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene    mutation source    AA range mutation size AA mutation type  \\\n",
       "0   b4484  pangenome variant  (120, 120)             1     substitution   \n",
       "0   b4484  pangenome variant  (105, 105)             1     substitution   \n",
       "0   b4484  pangenome variant  (164, 164)             1     substitution   \n",
       "0   b4484  pangenome variant    (11, 11)             1     substitution   \n",
       "0   b4484  pangenome variant  (147, 147)             1     substitution   \n",
       "..    ...                ...         ...           ...              ...   \n",
       "0   b4484  pangenome variant  (125, 125)             1     substitution   \n",
       "0   b4484  pangenome variant  (153, 153)             1     substitution   \n",
       "0   b4484  pangenome variant    (27, 27)             1     substitution   \n",
       "0   b4484  pangenome variant    (28, 28)             1     substitution   \n",
       "0   b4484  pangenome variant    (31, 31)             1     substitution   \n",
       "\n",
       "   AA ref seq AA seq change  \n",
       "0           M             T  \n",
       "0           E             D  \n",
       "0           R             H  \n",
       "0           S             L  \n",
       "0           T             A  \n",
       "..        ...           ...  \n",
       "0           T             K  \n",
       "0           S             T  \n",
       "0           G             V  \n",
       "0           D             Y  \n",
       "0           H             Y  \n",
       "\n",
       "[145 rows x 7 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "\n",
    "gene_var_df = pd.DataFrame()\n",
    "\n",
    "for blast_output_file in blast_inputs_mapping['blast output']:\n",
    "    blast_output_file_path = './data/blast_output/' + blast_output_file\n",
    "    gene = blast_output_file.replace('blast_','').replace('_alleles.xml','')\n",
    "    \n",
    "    for record in NCBIXML.parse(open(blast_output_file_path)):\n",
    "        if len(record.alignments) > 0:\n",
    "            # Description of available members: https://biopython.org/docs/1.75/api/Bio.Blast.Record.html\n",
    "            subject_match_start_pos = record.alignments[0].hsps[0].sbjct_start\n",
    "            blast_subject_str = record.alignments[0].hsps[0].sbjct\n",
    "            blast_match_str = record.alignments[0].hsps[0].match\n",
    "            blast_query_str = record.alignments[0].hsps[0].query\n",
    "            \n",
    "            # Won't find differences in the beginning of genes since doesn't compare query vs subject start and end match positions.\n",
    "            ins_d = _get_ins_d(blast_subject_str, blast_query_str, subject_match_start_pos)\n",
    "            for cnsc_pos_l in _get_consecutive_pos_l(ins_d.keys()):\n",
    "                seq_chng_str = ''.join([ins_d[p] for p in cnsc_pos_l])\n",
    "                df = pd.DataFrame.from_dict({\n",
    "                    'gene': gene,\n",
    "                    \"mutation source\": \"pangenome variant\",\n",
    "                    \"AA range\": (min(cnsc_pos_l),max(cnsc_pos_l)),\n",
    "                    \"mutation size\": len(cnsc_pos_l),\n",
    "                    \"AA mutation type\": \"insertion\",\n",
    "                    'AA ref seq': '',\n",
    "                    \"AA seq change\": seq_chng_str,\n",
    "                }, orient='index')  # Can't use columns due to diff parsing of 'AA range' input\n",
    "                df = df.T\n",
    "                gene_var_df = pd.concat([df, gene_var_df]) # Accounting for multiples of same allele\n",
    "            \n",
    "            # Won't find differences in the beginning of genes since doesn't compare query vs subject start and end match positions.\n",
    "            del_d = _get_del_d(blast_subject_str, blast_query_str, subject_match_start_pos)\n",
    "            for cnsc_pos_l in _get_consecutive_pos_l(del_d.keys()):\n",
    "                seq_chng_str = ''.join([del_d[p] for p in cnsc_pos_l])\n",
    "                df = pd.DataFrame.from_dict({\n",
    "                    'gene': gene,\n",
    "                    \"mutation source\": \"pangenome variant\",\n",
    "                    \"AA range\": (min(cnsc_pos_l),max(cnsc_pos_l)),\n",
    "                    \"mutation size\": len(cnsc_pos_l),\n",
    "                    \"AA mutation type\": \"deletion\",\n",
    "                    'AA ref seq': seq_chng_str,\n",
    "                    \"AA seq change\": '',\n",
    "                }, orient='index')  # Can't use columns due to diff parsing of 'AA range' input\n",
    "                df = df.T\n",
    "                gene_var_df = pd.concat([df, gene_var_df]) # Accounting for multiples of same allele\n",
    "                \n",
    "            sub_d = _get_sub_aa_d(\n",
    "                blast_subject_str,\n",
    "                blast_match_str,\n",
    "                blast_query_str,\n",
    "                list(ins_d.keys()) + list(del_d.keys()),\n",
    "                subject_match_start_pos)\n",
    "            \n",
    "            for cnsc_pos_l in _get_consecutive_pos_l(sub_d.keys()):\n",
    "                seq_ref_str = ''.join([sub_d[p]['s'] for p in cnsc_pos_l])\n",
    "                seq_chng_str = ''.join([sub_d[p]['q'] for p in cnsc_pos_l])\n",
    "                df = pd.DataFrame.from_dict({\n",
    "                    'gene': gene,\n",
    "                    \"mutation source\": \"pangenome variant\",\n",
    "                    \"AA range\": (min(cnsc_pos_l),max(cnsc_pos_l)),\n",
    "                    \"mutation size\": len(cnsc_pos_l),\n",
    "                    \"AA mutation type\": \"substitution\",\n",
    "                    'AA ref seq': seq_ref_str,\n",
    "                    \"AA seq change\": seq_chng_str,\n",
    "                }, orient='index')  # Can't use columns due to diff parsing of 'AA range' input\n",
    "                df = df.T\n",
    "                gene_var_df = pd.concat([df, gene_var_df]) # Accounting for multiples of same allele\n",
    "\n",
    "gene_var_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
